rclcpp/src/rclcpp/executors/executor_notify_waitable.cpp

```diff
@@ -47,21 +68,45 @@ ExecutorNotifyWaitable::add_to_wait_set(rcl_wait_set_t & wait_set)
 {
   std::lock_guard<std::mutex> lock(guard_condition_mutex_);
 
+  idxs_of_added_guard_condition_.clear();
+  idxs_of_added_guard_condition_.reserve(notify_guard_conditions_.size());
+
+  if(needs_processing) {
+    rcl_guard_condition_t * cond = &guard_condition_->get_rcl_guard_condition();
+    size_t rcl_index;
+    rcl_ret_t ret = rcl_wait_set_add_guard_condition(&wait_set, cond, &rcl_index);
+
+    if (RCL_RET_OK != ret) {
+      rclcpp::exceptions::throw_from_rcl_error(
+        ret, "failed to add guard condition to wait set");
+    }
+
+    idxs_of_added_guard_condition_.push_back(rcl_index);
+
+    // we want to directly wake up any way, not need to add the other guard conditions
+    guard_condition_->trigger();
+
+    return;
+  }
+
   // Note: no guard conditions need to be re-triggered, since the guard
   // conditions in this class are not tracking a stateful condition, but instead
   // only serve to interrupt the wait set when new information is available to
   // consider.
-  for (auto weak_guard_condition : this->notify_guard_conditions_) {
-    auto guard_condition = weak_guard_condition.lock();
+  for (const auto & guard_holder : this->notify_guard_conditions_) {
+    const auto & guard_condition = guard_holder.strong_reference;
     if (!guard_condition) {continue;}
```

Comment:
This doesn't seem right to me, are we not holding the strong_reference unconditionally? And if not should we not check if the weak one locks?

And it seems like we only use the weak_reference for comparison, then why not just have the strong reference and not have the GuardHolder at all?

Reply:
You are right, if we hold the shared_ptr, the lock of the weak_ptr will always work, and we don't need to do all of this, I will adjust the code.

Reply:
removed the holder stuff.